### PR TITLE
Add an EXCLUDES_LIST param to job generators

### PR DESCRIPTION
**Pull request by M-Davies**
Adds a parameter that allows a user to exclude certain configurations from being regenerated

Closes: #2045 
Signed-off-by: Morgan Davies <morgandavies2020@gmail.com>

### CODE REVIEW

**Comment by M-Davies** [comment]
Tested this with a test generator
- https://ci.adoptopenjdk.net/job/build-scripts/job/utils/job/alpha_pipeline_jobs_generator/17/ tests `EXCLUDES_LIST` populated with `{    "x64Mac": [        "hotspot",        "openj9"    ] }`. These platforms were not regenerated as expected
```
10:18:13  [INFO] Regenerating: x64Mac
[Pipeline] echo
10:18:13  [INFO] Regenerating variant x64Mac: hotspot...
[Pipeline] echo
10:18:13  [INFO] FOUND MATCH! buildConfiguration key: x64Mac and config file key: x64Mac
[Pipeline] echo
10:18:13  [INFO] Excluding mac: hotspot from jdk8u regeneration due to it being in the EXCLUDES_LIST...
[Pipeline] echo
10:18:13  [INFO] Regenerating variant x64Mac: openj9...
[Pipeline] echo
10:18:13  [INFO] FOUND MATCH! buildConfiguration key: x64Mac and config file key: x64Mac
[Pipeline] echo
10:18:13  [INFO] Excluding mac: openj9 from jdk8u regeneration due to it being in the EXCLUDES_LIST...
```
- https://ci.adoptopenjdk.net/job/build-scripts/job/utils/job/alpha_pipeline_jobs_generator/18/console tests `EXCLUDES_LIST` populated with `{    "x64MacXL": [       "openj9"    ] }`. This platform also was not regenerated as expected:
```
10:26:10  [INFO] Regenerating: x64MacXL
[Pipeline] echo
10:26:10  [INFO] Regenerating variant x64MacXL: openj9...
[Pipeline] echo
10:26:10  [INFO] FOUND MATCH! buildConfiguration key: x64MacXL and config file key: x64MacXL
[Pipeline] echo
10:26:10  [INFO] Excluding mac: openj9 from jdk8u regeneration due to it being in the EXCLUDES_LIST...
```
This is ready for review

**Comment by M-Davies** [comment]
@AdamBrousseau @andrew-m-leonard Could I get reviews please?

**Review by AdamBrousseau — COMMENTED** [review]
@M-Davies This is great and will be very useful. However I think it only addresses one part of #2045. If I wanted to generate with a custom config, for example, different node labels or different test suits, I would still have to fork/modify to get what I want. Unless I am missing another feature?

**Comment by M-Davies** [comment]
@AdamBrousseau See https://github.com/AdoptOpenJDK/openjdk-build/issues/2045#issuecomment-685617366. Those are the two use cases I identified that I thought would be most appropriate to resolve your issue. Short of forking/branching your own config files, I'm not sure how else you could keep in sync. Unless I'm missing something or misunderstanding your issue?

**Comment by M-Davies** [comment]
run tests

**Comment by M-Davies** [comment]
run tests

**Comment by M-Davies** [comment]
@AdamBrousseau Could I get your review please?

**Review by AdamBrousseau — APPROVED** [review]
LFTM. Thanks Morgan.

File: pipelines/build/common/config_regeneration.groovy [hunk 1]
```diff
@@ -1,6 +1,6 @@
 @Library('local-lib@master')
 import common.IndividualBuildConfig
-import groovy.json.JsonSlurper
+import groovy.json.*
```
Comment (karianna) [inline pipelines/build/common/config_regeneration.groovy:3]: Prefer explicit imports

File: pipelines/build/common/config_regeneration.groovy [hunk 2]
```diff
@@ -181,6 +214,13 @@ class Regeneration implements Serializable {
     */
     IndividualBuildConfig buildConfiguration(Map<String, ?> platformConfig, String variant, String javaToBuild) {
         try {
+
+            // Check if it's in the excludes list
+            if (overridePlatform(platformConfig, variant)) {
+                context.println "[INFO] Excluding $platformConfig.os: $variant from $javaToBuild regeneration due to it being in the EXCLUDES_LIST..."
+                return "EXCLUDED"
```
Comment (karianna) [inline pipelines/build/common/config_regeneration.groovy:221]: Magic string, declare as a const?

**Review by karianna — CHANGES_REQUESTED** [review]
Couple of changs - also does this need to go in before next week's release?  Else it'll have to wait.

**Comment by M-Davies** [comment]
> Couple of changs - also does this need to go in before next week's release?  Else it'll have to wait.

Unless @AdamBrousseau needs to it be in for the OpenJ9 release, it can wait until afterwards

**Comment by AdamBrousseau** [comment]
No rush on my side.

**Comment by karianna** [comment]
@M-Davies I think the comments also need addressing?

**Comment by M-Davies** [comment]
@karianna They've been addressed. Just waiting for https://ci.adoptopenjdk.net/job/build-scripts/job/utils/job/alpha_pipeline_jobs_generator/49/console to finish testing the changes to confirm everyone still works ok. Once it finishes (and hopefully passes), I could use a re-review

**Review by karianna — APPROVED** [review]

**Review by tellison — APPROVED** [review]

**Review by gdams — APPROVED** [review]

**Pull request merged**